enhancements/20240805-encrypted-block-volume.md

```diff
@@ -0,0 +1,440 @@
+
+Admin wants to encrypt the current boot volume for security, so Admin follows up the guideline:
+
+1. Prepare a virtual machine image, which you could upload or use URL to download.
+2. Create a secret for encryption and template storage class.
+3. Create a new virtual machine image with encryption enabled.
+
+After that, Admin gets an encrypted virtual machine image as boot volume. Then, Admin can start the virtual machine with selecting encrypted boot volume in GUI. 
+
+For security reason, Admin decides to delete that original virtual machine image. The encrypted image is independent of the original image. 
+
+For this moment, Admin only has this encrypted virtual machine image and Admin can reuse this boot volume.
+
+Over time, Admin wants to decrypt the encrypted virtual machine image. Admin follows up the guideline:
+
+1. Prepare an encrypted virtual machine image.
+2. Create a new virtual machine image with decryption enabled.
+
+After that, Admin will have an encrypted virtual machine image and an unencrypted virtual machine image. Admin could select one of them to start the virtual machine in GUI.
+
+
+#### Story 2 - Use encrypted data volume in virtual machine
+
+Admin wants to encrypt a data volume for security, so Admin follows up the guideline:
+
+1. Create a secret and template storage class for encryption/decryption.
+2. Select previous storage class when creating a new data volume.
+
+Now, Admin can read the data in the VM. Besides, Admin could attach this encrypted volume to any virtual machine, even as hot plug volume.
+
+
+### User Experience In Detail
+
+
+#### Encrypt a Virtual Machine Image
+
+
+1. Prepare a image
+
+2. Create a Secret
+    ```yaml
+    apiVersion: v1
+    kind: Secret
+    metadata:
+        name: longhorn-crypto
+        namespace: longhorn-system # you could create secret in any namespace
+    stringData:
+        CRYPTO_KEY_VALUE: "Some secret value here"
+        CRYPTO_KEY_PROVIDER: "secret"
+        CRYPTO_KEY_CIPHER: "aes-xts-plain64"
+        CRYPTO_KEY_HASH: "sha256"
+        CRYPTO_KEY_SIZE: "256"
+        CRYPTO_PBKDF: "argon2i"
+    ```
+3. Create Storage Class
+    ```yaml
+    kind: StorageClass
+    apiVersion: storage.k8s.io/v1
+    metadata:
+      name: harvester-longhorn-encryption
+    provisioner: driver.longhorn.io
+    allowVolumeExpansion: true
+    parameters:
+        numberOfReplicas: "2"
+        staleReplicaTimeout: "2880" # 48 hours in minutes
+        fromBackup: ""
+        migratable: "true"
+        encrypted: "true"
+        csi.storage.k8s.io/provisioner-secret-name: "longhorn-crypto"
+        csi.storage.k8s.io/provisioner-secret-namespace: "longhorn-system"
+        csi.storage.k8s.io/node-publish-secret-name: "longhorn-crypto"
+        csi.storage.k8s.io/node-publish-secret-namespace: "longhorn-system"
+        csi.storage.k8s.io/node-stage-secret-name: "longhorn-crypto"
+        csi.storage.k8s.io/node-stage-secret-namespace: "longhorn-system"
+    ```
```

Comment:
IIUC, the sc is not required for image encryption? Just want to make sure, thanks.

Reply:
In the first generation of solution, I bound the secret with `virtualmachineimage`. At that moment, we actually didn't need to create SC first. 

However, after our discussion, I think we could directly use storage class from annotation to pass secret to `backingimage`. It makes `virtualmachineimage` not to bind with secret, but with SC. I think it makes more sense for our current flow and more flexible.